### PR TITLE
Isis telemetry

### PR DESCRIPTION
**Pull request by SuryyaKrJana**
[https://github.com/open-traffic-generator/models-yang/blob/isis-telemetry/artifacts/open-traffic-generator-isis.txt](https://github.com/open-traffic-generator/models-yang/blob/isis-telemetry/artifacts/open-traffic-generator-isis.txt
)
References:
1. https://github.com/openconfig/public/blob/master/release/models/isis/openconfig-isis-lsp.yang
2. https://keysighttech.sharepoint.com/:u:/r/sites/Athena-G/Shared%20Documents/docs/openconfig-yang-tree/[openconfig_yang_tree_full.html](https://keysighttech.sharepoint.com/:u:/r/sites/Athena-G/Shared%20Documents/docs/openconfig-yang-tree/openconfig_yang_tree_full.html?csf=1&web=1&e=3LBg1n)?csf=1&web=1&e=3LBg1n
3. [RT-2] IS-IS
RT-2.1: Base IS-IS Process and Adjacencies
●	With a known LSP content, ensure that the telemetry received from the device for the LSP matches the expected content.

### CODE REVIEW

File: models/isis/open-traffic-generator-isis.yang [hunk 1]
```diff
@@ -107,566 +107,571 @@ module open-traffic-generator-isis {
...
-        "A reference to the Link State PDU ID.";
-      }
-
-      container state {
+        "Test.";
+      list lsps {
+        key "system-id lsp-index";
         description
-          "State parameters of Link State PDU.";
+          "The system ID of the router.";
```
Comment (ajbalogh) [inline models/isis/open-traffic-generator-isis.yang:116]: The description is for the key and should reflect that the key is a combination of the named keys

Comment (SuryyaKrJana) [reply models/isis/open-traffic-generator-isis.yang:121]: We have changed the key to be lsp-id.

File: artifacts/open-traffic-generator-isis.txt [hunk 2]
```diff
@@ -3,33 +3,124 @@ module: open-traffic-generator-isis
...
+           |     +--ro in-bcast-hellos?    otg-types:counter64
+           |     +--ro out-p2p-hellos?     otg-types:counter64
+           |     +--ro in-p2p-hellos?      otg-types:counter64
+           |     +--ro out-psnp?           otg-types:counter64
+           |     +--ro in-psnp?            otg-types:counter64
+           |     +--ro out-csnp?           otg-types:counter64
+           |     +--ro in-csnp?            otg-types:counter64
+           |     +--ro out-lsp?            otg-types:counter64
+           |     +--ro in-lsp?             otg-types:counter64
+           +--ro link-state-databases
```
Comment (PrasenjitAdhikary) [inline artifacts/open-traffic-generator-isis.txt:36]: For BGP, learned info is intentionally made parallel to the state, so that state info can be retrieved without forcing to fetch learned info. Otherwise, if user is only interested to fetch BGP stats, gNMI server always need to make two call, one for stats and other for states (learned info).

File: artifacts/open-traffic-generator-bgp.txt [hunk 3]
```diff
@@ -20,40 +20,28 @@ module: open-traffic-generator-bgp
...
         |     +--ro origin           -> ../state/origin
+        |     +--ro path-id          -> ../state/path-id
         |     +--ro state
         |        +--ro address?                 otg-types:ipv4-address
         |        +--ro prefix-length?           uint32
         |        +--ro origin?                  enumeration
         |        +--ro path-id?                 uint32
         |        +--ro next-hop-ipv4-address?   otg-types:ipv4-address
         |        +--ro next-hop-ipv6-address?   otg-types:ipv6-address
-        |        +--ro as-path* []
```
Comment (PrasenjitAdhikary) [inline artifacts/open-traffic-generator-bgp.txt:34]: This PR should focus only on ISIS. BGP changes should not be part of this PR

Comment (SuryyaKrJana) [reply artifacts/open-traffic-generator-bgp.txt:34]: not sure how this change has come.

File: models/isis/open-traffic-generator-isis.yang [hunk 4]
```diff
@@ -97,6 +105,442 @@ module open-traffic-generator-isis {
...
+  grouping lsp-states {
+    description
+      "This grouping defines ISIS LSP state information.";
+    reference
+      "1. https://github.com/openconfig/public/blob/master/release/models/isis/openconfig-isis-lsp.yang
+       2. http://ops.openconfig.net/branches/models/master/docs/openconfig-isis.html
+       3. https://github.com/openconfig/public/tree/master/release/models/isis";
+    container lsp-states {
+      description
+        "Test.";
```
Comment (PrasenjitAdhikary) [inline models/isis/open-traffic-generator-isis.yang:117]: Looks like a placeholder description.

File: models/isis/open-traffic-generator-isis.yang [hunk 5]
```diff
@@ -97,6 +105,442 @@ module open-traffic-generator-isis {
...
+      list lsps {
+        key "lsp-id";
+        description
+          "The system ID of the router.";
```
Comment (PrasenjitAdhikary) [inline models/isis/open-traffic-generator-isis.yang:121]: This description is for LSPs ?

File: artifacts/open-traffic-generator-isis.txt [hunk 6]
```diff
@@ -3,33 +3,124 @@ module: open-traffic-generator-isis
...
+                    |  +--ro flags*                enumeration
+                    |  +--ro is-type               uint8
+                    +--ro tlvs
+                       +--ro tlv* [type]
+                          +--ro type                          -> ../state/type
+                          +--ro state
+                          |  +--ro type    uint8
+                          +--ro hostname
+                          |  +--ro state
+                          |     +--ro hostname*   string
```
Comment (PrasenjitAdhikary) [inline artifacts/open-traffic-generator-isis.txt:55]: OTG model shows a single hostname, whereas here it is a list of hostnames.

File: artifacts/open-traffic-generator-isis.txt [hunk 7]
```diff
@@ -3,33 +3,124 @@ module: open-traffic-generator-isis
...
+                    |  +--ro sequence-number?      uint32
+                    |  +--ro pdu-length            uint16
+                    |  +--ro flags*                enumeration
+                    |  +--ro is-type               uint8
+                    +--ro tlvs
+                       +--ro tlv* [type]
+                          +--ro type                          -> ../state/type
+                          +--ro state
+                          |  +--ro type    uint8
+                          +--ro hostname
```
Comment (PrasenjitAdhikary) [inline artifacts/open-traffic-generator-isis.txt:53]: hostname or hostnames? If the child is a list, it should be hostnames

File: artifacts/open-traffic-generator-isis.txt [hunk 8]
```diff
@@ -3,33 +3,124 @@ module: open-traffic-generator-isis
...
+                          |        +--ro state
+                          |           +--ro system-id    otg-types:hex-string
+                          +--ro ipv4-internal-reachability
+                          |  +--ro prefixes
+                          |     +--ro prefix* [prefix]
+                          |        +--ro prefix            -> ../state/prefix
+                          |        +--ro state
+                          |        |  +--ro up-down?   boolean
+                          |        |  +--ro prefix?    otg-types:ipv4-address
+                          |        +--ro default-metric
```
Comment (PrasenjitAdhikary) [inline artifacts/open-traffic-generator-isis.txt:69]: Not matching with the OTG model. Only up/down state is available. Also the name is inconsistent with OTG model, where the attribute name is redistribution_type

File: artifacts/open-traffic-generator-isis.txt [hunk 9]
```diff
@@ -3,33 +3,124 @@ module: open-traffic-generator-isis
...
+                          |              +--ro flags?    enumeration
+                          |              +--ro metric?   uint8
+                          +--ro ipv4-external-reachability
+                          |  +--ro prefixes
+                          |     +--ro prefix* [prefix]
+                          |        +--ro prefix            -> ../state/prefix
+                          |        +--ro state
+                          |        |  +--ro up-down?   boolean
+                          |        |  +--ro prefix?    otg-types:ipv4-address
+                          |        +--ro default-metric
```
Comment (PrasenjitAdhikary) [inline artifacts/open-traffic-generator-isis.txt:80]: similar to ipv4-internal-reachability, only up/down state is available in the OTG model

File: artifacts/open-traffic-generator-isis.txt [hunk 10]
```diff
@@ -3,33 +3,124 @@ module: open-traffic-generator-isis
...
+                          |  +--ro prefixes
+                          |     +--ro prefix* [prefix]
+                          |        +--ro prefix     -> ../state/prefix
+                          |        +--ro state
+                          |        |  +--ro up-down?   boolean
+                          |        |  +--ro s-bit?     boolean
+                          |        |  +--ro prefix?    otg-types:ipv4-address
+                          |        |  +--ro metric?    uint32
+                          |        +--ro subtlvs
+                          |           +--ro subtlv* [type]
```
Comment (PrasenjitAdhikary) [inline artifacts/open-traffic-generator-isis.txt:100]: If we keep subtlv in yang-models, then it also has to be added in OTG-models.

File: artifacts/open-traffic-generator-isis.txt [hunk 11]
```diff
@@ -3,33 +3,124 @@ module: open-traffic-generator-isis
...
+                             +--ro prefixes
+                                +--ro prefix* [prefix]
+                                   +--ro prefix     -> ../state/prefix
+                                   +--ro state
+                                   |  +--ro up-down?   boolean
+                                   |  +--ro x-bit?     boolean
+                                   |  +--ro s-bit?     boolean
+                                   |  +--ro prefix?    otg-types:ipv6-address
+                                   |  +--ro metric?    uint32
+                                   +--ro subtlvs
```
Comment (PrasenjitAdhikary) [inline artifacts/open-traffic-generator-isis.txt:118]: subtlvs not present in OTG models

**Review by apratimmukherjee — APPROVED** [review]
One minor comment remains : databases -> database.

File: models/isis/open-traffic-generator-isis.yang [hunk 12]
```diff
@@ -97,6 +105,470 @@ module open-traffic-generator-isis {
...
+
+      leaf system-id {
+        type otg-types:hex-string;
+        mandatory true;
+        description
+          "System-ID of IS neighbor, e.g. '640100010000'.";
+      }
+    }
+
+    //uses isis-lsdb-common-metric-specification;
```
Comment (PrasenjitAdhikary) [inline models/isis/open-traffic-generator-isis.yang:469]: Commented code should be cleaned-up

File: models/isis/open-traffic-generator-isis.yang [hunk 13]
```diff
@@ -97,6 +105,470 @@ module open-traffic-generator-isis {
...
+    container state {
+     description
+       "State parameters of IPv4 standard prefix.";
+      leaf prefix {
+        type otg-types:ipv4-address;
+        description
+          "IPv4 prefix contained within reachability TLVs.";
+      }
+
+      leaf prefix-lenght {
```
Comment (PrasenjitAdhikary) [inline models/isis/open-traffic-generator-isis.yang:525]: typo in the property name.

**Review by PrasenjitAdhikary — APPROVED** [review]
One typo needs to be corrected. Otherwise LGTM.

File: models/isis/open-traffic-generator-isis.yang [hunk 14]
```diff
@@ -97,6 +105,442 @@ module open-traffic-generator-isis {
...
+                22.";
+
+              uses isis-lsdb-common-extisreach-neighbors;
+            }
+
+            container extended-ipv4-reachability {
+              description
+                "This container defines list of IPv4 extended reachability
+                information.
+                It is advertised when the 'wide metric' is enabled";
```
[the remark this answers is not in the document]
Comment (SuryyaKrJana) [reply models/isis/open-traffic-generator-isis.yang]: not relevent

**Comment by SuryyaKrJana** [comment]
Merging ISIS Telemetry to main
